### PR TITLE
Modify pipelines to get purls from package_data

### PR DESCRIPTION
**Pull request by AyanSinhaMahapatra**
The `scan_codebase_packages` pipeline currently takes a lot of time for large projects with ~700k resources, and it is stuck in the `scan_for_application_packages` step, and so presumably it's the package assembly in there that takes the longest.
The main objective of this pipeline was as a precursor to the `populate_purldb` pipeline to get purls from the codebase and submit them to purldb for indexing. So this PR does the following:

1. Disables the package assembly for the `scan_codebase_packages` pipeline. So we only have the codebase scanned for package_data and these are added to the specific resources without creating DiscoveredPackage and DiscoveredDependency objects out of these package data, by assembling.
2. If there are no packages/dependencies added to the project (otherwise we will add purls from these packages/deps twice), then further look for package data detected in the resources and add purls from those.

### CODE REVIEW

File: scanpipe/pipes/scancode.py [hunk 1]
```diff
@@ -430,6 +431,20 @@ def assemble_packages(project):
                     logger.info(f"Unknown Package assembly item type: {item!r}")
 
 
+def get_packages_with_purl_from_resources(project):
```
Comment (tdruez) [inline scanpipe/pipes/scancode.py:434]: Missing a docstring and unit tests.

File: scanpipe/pipelines/populate_purldb.py [hunk 2]
```diff
@@ -48,6 +50,18 @@ def populate_purldb_with_discovered_dependencies(self):
             package_type="DiscoveredDependency",
         )
 
+    def populate_purldb_with_detected_purls(self):
+        """Add DiscoveredPackage to PurlDB."""
+        if (
+            not self.project.discoveredpackages.exists()
+            and not self.project.discovereddependencies.exists()
+        ):
```
Comment (tdruez) [inline scanpipe/pipelines/populate_purldb.py:58]: You can wrap the condition into a explicit variable to help the readability.
```suggestion
        no_packages_and_no_dependencies = all([
            not self.project.discoveredpackages.exists(),
            not self.project.discovereddependencies.exists(),
        ])
        
        if no_packages_and_no_dependencies:
```

File: scanpipe/pipelines/populate_purldb.py [hunk 3]
```diff
@@ -48,6 +50,18 @@ def populate_purldb_with_discovered_dependencies(self):
             package_type="DiscoveredDependency",
         )
 
+    def populate_purldb_with_detected_purls(self):
```
Comment (tdruez) [inline scanpipe/pipelines/populate_purldb.py:53]: It's unclear why this is only called when the project has not packages/dependencies.

**Review by tdruez — CHANGES_REQUESTED** [review]
See various comments.

Also: 
> Disables the package assembly for the scan_codebase_packages pipeline. So we only have the codebase scanned for package_data and these are added to the specific resources without creating DiscoveredPackage and DiscoveredDependency objects out of these package data, by assembling.

I'm not sure about changing the behavior of package detection in `scan_codebase_packages`. It's unclear to me looking at the code the consequence of this.

File: scanpipe/pipelines/scan_codebase_packages.py [hunk 4]
```diff
@@ -36,3 +37,7 @@ def steps(cls):
             cls.flag_ignored_resources,
             cls.scan_for_application_packages,
         )
+
+    def scan_for_application_packages(self):
+        """Scan unknown resources for packages information."""
+        scancode.scan_for_application_packages(self.project, assemble=False)
```
[the remark this answers is not in the document]
Comment (AyanSinhaMahapatra) [reply scanpipe/pipelines/scan_codebase_packages.py:46]: Added this, thanks!

File: scanpipe/pipes/scancode.py [hunk 5]
```diff
@@ -338,7 +338,7 @@ def scan_for_files(project, resource_qs=None):
     _scan_and_save(resource_qs, scan_file, save_scan_file_results, scan_func_kwargs)
 
 
-def scan_for_application_packages(project):
+def scan_for_application_packages(project, assemble=True):
     """
     Run a package scan on files without a status for a `project`,
     then create DiscoveredPackage and DiscoveredDependency instances
```
[the remark this answers is not in the document]
Comment (AyanSinhaMahapatra) [reply scanpipe/pipes/scancode.py]: Added some details on this, please review if this makes sense now. Thanks!

File: scanpipe/pipes/scancode.py (hunk 1 above)
Comment (AyanSinhaMahapatra) [reply scanpipe/pipes/scancode.py:439]: Added a docstring and a test, thanks!

File: scanpipe/pipelines/populate_purldb.py (hunk 2 above)
Comment (AyanSinhaMahapatra) [reply scanpipe/pipelines/populate_purldb.py]: Added this, thanks!

File: scanpipe/pipelines/populate_purldb.py (hunk 3 above)
Comment (AyanSinhaMahapatra) [reply scanpipe/pipelines/populate_purldb.py:53]: This step exists to get purls out of package_data saved at resources. These are still detected and saved, even though we don't have Package/Dependency instances (the assemble=False skips the package assembly, but resource level package detection is still enabled and happening).

**Comment by AyanSinhaMahapatra** [comment]
> I'm not sure about changing the behavior of package detection in scan_codebase_packages. It's unclear to me looking at the code the consequence of this.

If the main objective of this pipeline is to get purls from the codebase and use this as a previous step for the populate_purldb pipeline, then we need to create Discoveredpackage/DiscoveredDependency instances, but having resource level package detections in each resource is enough (We can goa  step further by passing an argument to scancode to only detect purl related information only, and ignore everything else, which will be the next step). The package assembly step uses these resource level detections and creates instances, assigns files, but this is not required in this context. 

@tdruez thanks for the review and feedback, I've addressed these, ready for your review again :smile:

**Review by tdruez — CHANGES_REQUESTED** [review]
@AyanSinhaMahapatra It would be nice to add some details in the docstring of ScanCodebasePackages to describe its purpose and its limitations.

File: scanpipe/pipelines/populate_purldb.py (hunk 3 above)
Comment (AyanSinhaMahapatra) [reply scanpipe/pipelines/populate_purldb.py:53]: @tdruez I've added this in the code as comments, thanks!

File: scanpipe/pipelines/populate_purldb.py [hunk 6]
```diff
@@ -70,7 +70,7 @@ def feed_purldb(self, packages, package_type):
         if not purldb.is_available():
             raise Exception("PurlDB is not available.")
 
-        package_urls = [package.purl for package in packages]
+        package_urls = list(set([package.purl for package in packages]))
```
Comment (AyanSinhaMahapatra) [inline scanpipe/pipelines/populate_purldb.py:77]: Note that when we get purls from package_data and not package/dependency instances, we might have duplicate purls (like from manifests of a same package instance and from different manifests having same dependencies), so it makes sense to send only the unique purls for indexing in purldb.

File: scanpipe/pipes/scancode.py [hunk 7]
```diff
@@ -436,6 +442,25 @@ def assemble_packages(project):
...
+        for package_mapping in resource.package_data:
+            for dep in package_mapping.get("dependencies"):
+                yield packagedcode_models.Dependency.from_dependent_package(
+                    dependent_package=dep,
+                    datafile_path=resource.path,
+                    datasource_id=package_mapping.get("datasource_id"),
+                    package_uid=None,
+                )
+            pd = packagedcode_models.PackageData.from_dict(mapping=package_mapping)
+            yield pd
```
Comment (tdruez) [inline scanpipe/pipes/scancode.py:461]: Assigning to a var is not useful here unless the var name helps with the readability, which is not the case.
```suggestion
            yield packagedcode_models.PackageData.from_dict(mapping=package_mapping)
```

File: scanpipe/pipes/scancode.py [hunk 8]
```diff
@@ -436,6 +442,25 @@ def assemble_packages(project):
...
+    """
+    Yield Dependency or PackageData objects created from detected package_data
+    in all the project resources. Both Dependency and PackageData objects have
+    the `purl` attribute with a valid purl.
+    """
+    for resource in project.codebaseresources.has_package_data():
+        for package_mapping in resource.package_data:
+            for dep in package_mapping.get("dependencies"):
+                yield packagedcode_models.Dependency.from_dependent_package(
+                    dependent_package=dep,
```
Comment (tdruez) [inline scanpipe/pipes/scancode.py:455]:
```suggestion
            for dependency in package_mapping.get("dependencies"):
                yield packagedcode_models.Dependency.from_dependent_package(
                    dependent_package=dependency,
```

**Review by tdruez — CHANGES_REQUESTED** [review]
2 more minor suggestions and we'll be ready ;)

File: scanpipe/pipes/scancode.py (hunk 7 above)
Comment (AyanSinhaMahapatra) [reply scanpipe/pipes/scancode.py]: Done, thanks!

File: scanpipe/pipes/scancode.py (hunk 8 above)
Comment (AyanSinhaMahapatra) [reply scanpipe/pipes/scancode.py]: Done, thanks!

**Comment by AyanSinhaMahapatra** [comment]
@tdruez latest comments addressed, thanks for the suggestions. :smile:

**Pull request merged**